library/common/http/dispatcher.cc

```diff
@@ -49,75 +65,152 @@ void Dispatcher::DirectStreamCallbacks::onHeaders(HeaderMapPtr&& headers, bool e
   default:
     error_code_ = ENVOY_UNDEFINED_ERROR;
   }
-  ENVOY_LOG(debug, "[S{}] intercepted local response", stream_handle_);
+  ENVOY_LOG(debug, "[S{}] intercepted local response", direct_stream_.stream_handle_);
   if (end_stream) {
-    // The local stream may or may not have completed. We don't want to be tracking/synchronized
-    // on that state, so we just reset everything now to ensure teardown.
-    auto stream = http_dispatcher_.getStream(stream_handle_);
-    ASSERT(stream);
-    stream->underlying_stream_.reset();
+    // The local stream may or may not have completed.
+    // If the local is not closed envoy will fire the reset for us.
+    // @see Dispatcher::DirectStreamCallbacks::closeRemote.
+    // Otherwise fire the reset from here.
+    if (direct_stream_.local_closed_) {
+      onReset();
+    }
   }
 }
 
-void Dispatcher::DirectStreamCallbacks::onData(Buffer::Instance& data, bool end_stream) {
-  ENVOY_LOG(debug, "[S{}] response data for stream (length={} end_stream={})", stream_handle_,
-            data.length(), end_stream);
+void Dispatcher::DirectStreamCallbacks::encodeData(Buffer::Instance& data, bool end_stream) {
+  ENVOY_LOG(debug, "[S{}] response data for stream (length={} end_stream={})",
+            direct_stream_.stream_handle_, data.length(), end_stream);
   if (!error_code_.has_value()) {
-    bridge_callbacks_.on_data(Buffer::Utility::toBridgeData(data), end_stream,
-                              bridge_callbacks_.context);
+    if (direct_stream_.dispatchable()) {
+      ENVOY_LOG(debug,
+                "[S{}] dispatching to platform response data for stream (length={} end_stream={})",
+                direct_stream_.stream_handle_, data.length(), end_stream);
+      bridge_callbacks_.on_data(Buffer::Utility::toBridgeData(data), end_stream,
+                                bridge_callbacks_.context);
+    }
+    closeRemote(end_stream);
   } else {
+    // FIXME: before merge. should this be an ASSERT? Yes, if we are in the local response path,
+    // then Envoy finishes the sequence with an encodeData. What if it later does trailers?
```

Comment:
Seems a little bit unlikely to me. But we could change this to a comment saying "This assert helps guard our reliance on current convention: an Envoy-generated error response always ends with a single data frame."

Reply:
from offline: leave the assert. But add explanation of why this assert is here.